doc/kernel/usermode/kernelobjects.rst

```diff
@@ -0,0 +1,226 @@
+
+Any memory reserved for a kernel object must be used exclusively for that
+object. Kernel objects may not be members of a union data type.
+
+The debug output of the ``gen_kobject_list.py`` script may be useful when
+debugging why some object was unexpectedly not being tracked. This
+information will be printed if the script is run with the ``--verbose`` flag,
+or if the build system is invoked with verbose output.
+
+Implementation Details
+----------------------
+
+The ``gen_kobject_list.py`` script is a post-build step which finds all the
+valid kernel object instances in the binary. It accomplishes this by parsing
+the DWARF debug information present in the generated ELF file for the kernel.
+
+Any instances of structs or arrays corresponding to kernel objects that meet
+the object placement criteria will have their memory addresses placed in a
+special perfect hash table of kernel objects generated by the 'gperf' tool.
+When a system call is made and the kernel is presented with a memory address
+of what may or may not be a valid kernel object, the address can be validated
+with a constant-time lookup in this table.
+
+Drivers are a special case. All drivers are instances of :c:type:`struct
+device`, but it is important to know what subsystem a driver belongs to so that
+incorrect operations, such as calling a UART API on a sensor driver object, can
+be prevented. When a device struct is found, its API pointer is examined to
+determine what subsystem the driver belongs to.
+
+The table itself maps kernel object memory addresses to instances of
+:c:type:`struct _k_object`, which has all the metadata for that object. This
+includes:
+
+* A bitfield indicating permissions on that object. All threads have a
+  numerical ID assigned to them at build time, used to index the permission
+  bitfield for an object to see if that thread has permission on it.
+* A type field indicating what kind of object this is, which is some
+  instance of :cpp:enum:`k_objects`.
+* A set of flags for that object. This is currently used to track
+  initialization state and whether an object is public or not.
+* An extra data field. This is currently used for thread stack objects
+  to denote how large the stack is, and for thread objects to indicate
+  the thread's index in kernel object permission bitfields.
+
+Supervisor Thread Access Permission
+===================================
+
+Supervisor threads can access any kernel object. However, permissions for
+supervisor threads are still tracked for two reasons:
+
+* If a supervisor thread calls :cpp:func:`k_thread_user_mode_enter()`, the
+  thread will then run in user mode with any permissions it had been granted
+  (in many cases, by itself) when it was a supervisor thread.
+
+* If a supervisor thread creates a user thread with the
+  :c:macro:`K_INHERIT_PERMS` option, the child thread will be granted the
+  same permissions as the parent thread, except the parent thread object.
```

Comment:
I'm not really clear on how a thread object is a permission; what am I missing? Does this mean it can access any objects accessible by the parent thread, except for that parent thread object itself?

Reply:
a struct k_thread is itself a kernel object. the child will not have permission inherited to make API calls on its own parent; if this is really desired permission must be granted explicitly.